format/leveldb/leveldb_table.go

```diff
@@ -0,0 +1,443 @@
+		// decompress if needed
+		d.SeekAbs(start)
+		if compressionType == compressionTypeNone {
+			d.FieldStruct("uncompressed", func(d *decode.D) {
+				readTableBlockContents(size, d)
+			})
+		} else {
+			compressedSize := size
+			compressed := data
+			bb := &bytes.Buffer{}
+			switch compressionType {
+			case compressionTypeSnappy:
+				decompressed, err := snappy.Decode(nil, compressed)
+				if err != nil {
+					d.Errorf("failed decompressing data: %v", err)
+				}
+				d.Copy(bb, bytes.NewReader(decompressed))
+			default:
+				d.Errorf("Unsupported compression type: %x", compressionType)
+			}
+			if bb.Len() > 0 {
+				d.FieldStructRootBitBufFn(
+					"uncompressed",
+					bitio.NewBitReader(bb.Bytes(), -1),
+					func(d *decode.D) {
+						readTableBlockContents(int64(bb.Len()), d)
+					},
+				)
+			}
+			d.FieldRawLen("compressed", compressedSize*8)
+		}
+
+	})
+}
+
+// Read content encoded as a sequence of key/value-entries and a trailer of restarts.
+// https://github.com/google/leveldb/blob/main/table/block_builder.cc#L16
+// https://github.com/google/leveldb/blob/main/table/block.cc#L48
+func readKeyValueContents(
+	keyCallbackFn func(sharedBytes []byte, unsharedSize int, d *decode.D) error,
+	valueCallbackFn func(d *decode.D),
+	size int64,
+	d *decode.D,
+) {
+	start := d.Pos()
+	end := start + size*8
+
+	var restartOffset int64
+	d.SeekAbs(end - uint32BitSize)
+	d.FieldStruct("trailer", func(d *decode.D) {
+		numRestarts := int64(d.FieldU32("num_restarts"))
+		restartOffset = size*8 - (1+numRestarts)*uint32BitSize
+		d.SeekAbs(start + restartOffset)
+		d.FieldArray("restarts", func(d *decode.D) {
+			for i := 0; i < int(numRestarts); i++ {
+				d.FieldU32("restart")
+			}
+		})
+	})
+	// TK: how do you make an empty entries-array appear _above_ the trailer?
+	// Right now, its omitted if empty.
+	if restartOffset <= 0 {
+		return
+	}
+	d.SeekAbs(start)
+	d.FieldArray("entries", func(d *decode.D) {
+		var lastKey []byte
+		for d.Pos() < start+restartOffset {
+			d.FieldStruct("entry", func(d *decode.D) {
+				// https://github.com/google/leveldb/blob/main/table/block.cc#L48-L75
+				shared := int64(d.FieldULEB128("shared_bytes"))
+				unshared := int64(d.FieldULEB128("unshared_bytes"))
+				valueLength := int64(d.FieldULEB128("value_length"))
+
+				// read key
+				// https://github.com/google/leveldb/blob/main/table/block.cc#L261
+				if int(shared) > len(lastKey) {
+					d.Fatalf("`shared` size is larger than `lastKey` length")
+				}
+				keyPrefix := lastKey[:shared]
+				keySuffix := readBytesWithoutChangingPosition(unshared, d)
+				lastKey = append(keyPrefix, keySuffix...)
+
+				if keyCallbackFn == nil && shared == 0 {
+					d.FieldUTF8("key", int(unshared))
+				} else {
+					d.FieldStruct("key", func(d *decode.D) {
+						if keyCallbackFn == nil {
+							keyCallbackFn = readPrefixedBytes
+						}
+						err := keyCallbackFn(keyPrefix, int(unshared), d)
+						if err != nil {
+							d.Errorf(err.Error())
```

Comment:
Maybe these should be `d.Errorf("%s", err)` to be sure some random `%s` etc in the error string is seen as a format?

Reply:
Thank you for the pointer. I just saw earlier that I'd used some `d.Errorf("%v", err)` code I had copy-pasted from avro_ocf.go without remembering.